### DISCUSSION TITLE
only small axons are detected on the notebook tutorial

### DISCUSSION THREAD

**Discussion by Mousa-saleh** [neuropoly/axondeepseg#532] [Show and tell]
Hello all,

I ran a test image on the getting started notebook.
It seems that the model only detect small neurons.
Can any help me figuring out what is the problem??

Thanks in advance,
![image](https://user-images.githubusercontent.com/85881282/122427631-f27a9d80-cf99-11eb-9405-fda0b60b00d6.png)

**Comment by vasudev-sharma** [comment]
Hey @Mousa-saleh thanks for opening a new discussion. 

Can you provide details which model you tried for your image? Also, do you have information about what's the modality of your data?

Assuming that you are not aware about modality of your data

As a first step, I suggest playing with different models, on the raw image, such as **SEM** (default_SEM_model), **TEM** (default_TEM_model), and **OM** (model_seg_pns_bf) model. By default, the notebook uses **SEM** model.  You can change the `model_name` in the notebook's cell accordingly.

**NOTE: -** Be sure, to change the `resampled_resolutions` parameter in the `axon_segmentation` function accordingly for these models.
a) For **SEM** and **model_seg_pns_bf**, set `resampled_resolutions=0.1`
<s>b) For **TEM** and **model_seg_pns_bf**, set `resampled_resolutions=0.001`</s>
b) For **TEM** , set `resampled_resolutions=0.01`

Depending on the results of the segmentations, you can decide what's the modality of your data is. 

---
Now, once you have identified the modality of your data, next step would be playing with the pixel size to get best segmentation.
The [`axon_segmentation`](https://github.com/neuropoly/axondeepseg/blob/6cc6ade5f8aa594a66ddec556f08ab15501d12d3/AxonDeepSeg/apply_model.py#L192) function in this notebook has a parameter named `acquired_resolution`, which is nothing but the pixel size of the image. I suggest at first set the `acquired_resolution` to "true pixel size" of your image. 

If optimal results are not obtained, try increasing or decreasing the `acquired_resolution` parameter incrementally and monitor the results of the segmentations.

You will notice that doing so you will find an optimal `acquired_resolution` parameter (pixel size) which works best for your case.

Let us know how the analysis goes.

**Reply by Mousa-saleh** [reply]
The data is collected using optical microscope so I used model_seg_pns_bf model.
I used resampled_resolution=0.1 as indicated in the documentation!!
I will change it to 0.001 and see what I get,
I will also try playing with the acquired resolution as you said.

Thanks @vasudev-sharma

**Reply by Mousa-saleh** [reply]
I used resampled_resolution=0.001 and got an out of memory error.
I increased it to 0.01 and got zero dice for axon and myelin, so I instead increased it and got the best value at 0.3.
I then started playing with acquired resoluton but got errors, the only case it worked when it equals half the resampled_resolution

I got the best dice values for axon(75%) and for myelin(67%) when using the following settings
resampled_resolution=0.3, acquired_resolution=0.15(the real value),
![image](https://user-images.githubusercontent.com/85881282/122472573-713aff00-cfc9-11eb-8f6a-d31480ff4df1.png)

when I set ,for example,  resampled_resolution=0.4 acquired_resolution=0.1, I get this error:

 ValueError                                Traceback (most recent call last)
<ipython-input-8-ceaa55cf9adf> in <module>
----> 1 prediction = axon_segmentation([path_testing], ["image.png"], path_model, config_network, resampled_resolutions=0.4,acquired_resolution=0.1, verbosity_level=3)

c:\users\lenovo\axondeepseg\AxonDeepSeg\apply_model.py in axon_segmentation(path_acquisitions_folders, acquisitions_filenames, path_model_folder, config_dict, ckpt_name, segmentations_filenames, inference_batch_size, overlap_value, resampled_resolutions, acquired_resolution, prediction_proba_activate, write_mode, gpu_per, verbosity_level)
    269                                    overlap_value=overlap_value, resampled_resolutions=resampled_resolutions,
    270                                    prediction_proba_activate=prediction_proba_activate, gpu_per=gpu_per,
--> 271                                    verbosity_level=verbosity_level)
    272         # Predictions are shape of image, value = class of pixel
    273 

c:\users\lenovo\axondeepseg\AxonDeepSeg\apply_model.py in apply_convnet(path_acquisitions, acquisitions_resolutions, path_model_folder, config_dict, ckpt_name, inference_batch_size, overlap_value, resampled_resolutions, prediction_proba_activate, gpu_per, verbosity_level)
    124             current_batch_prediction = perform_batch_inference(model, sess, pred, x, batch_x, inference_batch_size,
    125                                                                patch_size, n_classes,
--> 126                                                                prediction_proba_activate=prediction_proba_activate)
    127             # Update of the predictions lists.
    128             predictions_list.extend(current_batch_prediction)

c:\users\lenovo\axondeepseg\AxonDeepSeg\apply_model.py in perform_batch_inference(model, tf_session, tf_prediction_op, tf_input, batch_x, size_batch, input_size, n_classes, prediction_proba_activate)
    473     """
    474 
--> 475     batch_x = np.reshape(batch_x,(size_batch, input_size, input_size, 1))
    476 
    477     p = model.predict(batch_x)

<__array_function__ internals> in reshape(*args, **kwargs)

~\miniconda3\envs\ads_venv\lib\site-packages\numpy\core\fromnumeric.py in reshape(a, newshape, order)
    297            [5, 6]])
    298     """
--> 299     return _wrapfunc(a, 'reshape', newshape, order=order)
    300 
    301 

~\miniconda3\envs\ads_venv\lib\site-packages\numpy\core\fromnumeric.py in _wrapfunc(obj, method, *args, **kwds)
     56 
     57     try:
---> 58         return bound(*args, **kwds)
     59     except TypeError:
     60         # A TypeError occurs if the object does have such a method in its

ValueError: cannot reshape array of size 44032 into shape (1,512,512,1)

**Reply by vasudev-sharma** [reply]
Sorry my bad, the **resampled_resolutions** for our dataset, **Optical microscopy (model_seg_pns_bf)** should be 0.1 and for **TEM** it should be 0.01; that's probably why you got "out of memory error".  I've modified my previous comment. Thanks for pointing that out.

The segmentation results are far better than what you shown earlier. Thanks for letting us know the experiments you did with **acquired_resolution** and **resampled_resolutions** parameter. 

The reason you encountered _ValueError_ is the that the `resampled_resolutions=0.4` it too high to resample the image into patch size of **512**. 

If there are some more pointers to improve the segmentations, perhaps @mariehbourget can nudge you in the right direction .

**Comment by mariehbourget** [comment]
Hi @Mousa-saleh, thank you for letting us know about your experiments, it is good to see that the model performs well on your data.

I think you are on the right track with increasing the `resampled_resolutions` and you would probably achieve an even better results with a higher value (and keeping `acquired_resolutions=0.15`). As @vasudev-sharma mentioned, the `Value Error` you got when doing that is due to the minimum patch size  of 512x512 after resampling.

But don't worry, there is a way to work around that. To be able to increase the `resampled_resolutions`, you will first need to increase the size of your images by padding with black pixels around it.

The resampling in ADS goes like this:
`dimension of the image (in pixels) * acquired_resolutions / resampled_resolutions` and the result must be greater than 512.

For example, with your image of around `1025x1375`, `acquired_resolutions=0.15` and `resampled_resolutions=0.4`, you will have the following result (on the smallest dimension):
`1025*0.15/0.4 = 384 < 512`

So, you need to pad and save a new image with a size around 1375 (or more) to have a result greater than 512.
`1375*0.15/0.4 = 515 > 512`

**Reply by Mousa-saleh** [reply]
Hey @mariehbourget, it worked.
Can I calculate the dice only for the original image size before padding, that would further enhance the results 
As can be seen in the following picture, part of the padding area is mislabeled which decreased the dice values.
![image](https://user-images.githubusercontent.com/85881282/122613536-af99f200-d08d-11eb-8d5d-b6b1d5912a5a.png)

**Reply by mariehbourget** [reply]
Hi @Mousa-saleh, that is good news, thanks for the update!
The restrictions on patch size and image size only apply at the segmentation step. Once the segmentation is done, you can crop the arrays of both your segmentation and ground truth at the original dimension to compute the dice.
